File: python_pure/tests/test_hedging.py
from __future__ import annotations

import pytest
from volarb_py.hedging import (
    BandPolicy,
    HedgingInputs,
    InvalidHedgingInputsError,
    hedging_statistics,
    whalley_wilmott_band,
)

BASE = HedgingInputs(
    spot=100.0,
    strike=100.0,
    years_to_expiry=0.25,
    volatility=0.20,
    steps=63,
    proportional_cost=0.0010,
    risk_aversion=0.10,
)
PATHS = 400
WIDTHS = (0.0, 0.05, 0.10, 0.20, 0.40)
TUNING_GRID = (0.0, 0.05, 0.10, 0.20, 0.30, 0.50)
HELD_OUT_SEQUENCES = (2, 3, 4, 5, 6, 7, 8, 9, 10, 11)
SURVIVING_EDGE_FRACTION = 5.0
REPLICATION_PROFIT = 0.1
REPLICATION_SPREAD = 0.3


def run(policy: BandPolicy, sequence: int) -> object:
    return hedging_statistics(BASE, policy, 1, sequence, PATHS)


def test_the_band_scales_as_the_cube_root_of_cost_and_gamma_squared() -> None:
    base = whalley_wilmott_band(100.0, 0.04, 0.001, 0.1)
    assert whalley_wilmott_band(100.0, 0.04, 0.008, 0.1) == pytest.approx(2.0 * base, rel=1e-12)
    assert whalley_wilmott_band(100.0, 0.32, 0.001, 0.1) == pytest.approx(4.0 * base, rel=1e-12)
    assert whalley_wilmott_band(100.0, 0.04, 0.001, 0.8) == pytest.approx(0.5 * base, rel=1e-12)
    assert whalley_wilmott_band(100.0, 0.0, 0.001, 0.1) == 0.0


def test_a_wider_band_trades_less_and_pays_less() -> None:
    costs = [run(BandPolicy("fixed", width), 1).mean_transaction_cost for width in WIDTHS]
    counts = [run(BandPolicy("fixed", width), 1).mean_rebalance_count for width in WIDTHS]
    assert costs == sorted(costs, reverse=True)
    assert counts == sorted(counts, reverse=True)


def test_a_wider_band_carries_more_risk() -> None:
    spreads = [run(BandPolicy("fixed", width), 1).profit_standard_deviation for width in WIDTHS]
    assert spreads == sorted(spreads)


def test_hedging_with_no_cost_and_no_band_replicates_the_option() -> None:
    inputs = HedgingInputs(100.0, 100.0, 0.25, 0.20, 252, 0.0, 0.10)
    stats = hedging_statistics(inputs, BandPolicy("fixed", 0.0), 1, 1, PATHS)
    assert stats.mean_transaction_cost == 0.0
    assert abs(stats.mean_profit) < REPLICATION_PROFIT
    assert stats.profit_standard_deviation < REPLICATION_SPREAD


def test_the_same_seed_reproduces_the_same_statistics() -> None:
    one = run(BandPolicy("whalley_wilmott", 0.0), 1)
    other = run(BandPolicy("whalley_wilmott", 0.0), 1)
    assert one == other
    assert run(BandPolicy("whalley_wilmott", 0.0), 2) != one


def test_the_certainty_equivalent_is_the_mean_penalised_by_variance() -> None:
    stats = run(BandPolicy("fixed", 0.10), 1)
    variance = stats.profit_standard_deviation**2
    assert stats.certainty_equivalent == pytest.approx(
        stats.mean_profit - 0.5 * BASE.risk_aversion * variance, rel=1e-12
    )
    assert stats.certainty_equivalent < stats.mean_profit


def tuned_fixed_width() -> float:
    return max(TUNING_GRID, key=lambda width: run(BandPolicy("fixed", width), 1).certainty_equivalent)


def test_a_tuned_fixed_band_wins_in_sample() -> None:
    tuned = tuned_fixed_width()
    assert (
        run(BandPolicy("fixed", tuned), 1).certainty_equivalent
        > run(BandPolicy("whalley_wilmott", 0.0), 1).certainty_equivalent
    )


def test_a_tuned_fixed_band_degrades_on_every_held_out_set() -> None:
    tuned = tuned_fixed_width()
    in_sample = run(BandPolicy("fixed", tuned), 1).certainty_equivalent
    for sequence in HELD_OUT_SEQUENCES:
        assert run(BandPolicy("fixed", tuned), sequence).certainty_equivalent < in_sample


def test_the_tuned_advantage_does_not_survive_out_of_sample() -> None:
    tuned = tuned_fixed_width()
    in_sample_edge = (
        run(BandPolicy("fixed", tuned), 1).certainty_equivalent
        - run(BandPolicy("whalley_wilmott", 0.0), 1).certainty_equivalent
    )
    held_out_edges = [
        run(BandPolicy("fixed", tuned), sequence).certainty_equivalent
        - run(BandPolicy("whalley_wilmott", 0.0), sequence).certainty_equivalent
        for sequence in HELD_OUT_SEQUENCES
    ]
    mean_edge = sum(held_out_edges) / len(held_out_edges)
    assert in_sample_edge > 0.0
    assert mean_edge < in_sample_edge / SURVIVING_EDGE_FRACTION


def test_malformed_hedging_inputs_are_rejected() -> None:
    policy = BandPolicy("fixed", 0.1)
    for broken in (
        HedgingInputs(-1.0, 100.0, 0.25, 0.20, 63, 0.001, 0.10),
        HedgingInputs(100.0, 0.0, 0.25, 0.20, 63, 0.001, 0.10),
        HedgingInputs(100.0, 100.0, 0.0, 0.20, 63, 0.001, 0.10),
        HedgingInputs(100.0, 100.0, 0.25, 0.0, 63, 0.001, 0.10),
        HedgingInputs(100.0, 100.0, 0.25, 0.20, 0, 0.001, 0.10),
        HedgingInputs(100.0, 100.0, 0.25, 0.20, 63, -0.001, 0.10),
        HedgingInputs(100.0, 100.0, 0.25, 0.20, 63, 0.001, 0.0),
    ):
        with pytest.raises(InvalidHedgingInputsError):
            hedging_statistics(broken, policy, 1, 1, PATHS)

    with pytest.raises(InvalidHedgingInputsError):
        hedging_statistics(BASE, BandPolicy("fixed", -1.0), 1, 1, PATHS)
    with pytest.raises(InvalidHedgingInputsError):
        hedging_statistics(BASE, policy, 1, 1, 1)

File: cpp_core/tests/test_hedging.cpp
#include "volarb/hedging.hpp"

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

#include <cmath>
#include <vector>

using Catch::Approx;
using volarb::BandPolicy;
using volarb::BandRule;
using volarb::band_rule_from_name;
using volarb::HedgingInputs;
using volarb::HedgingStatistics;
using volarb::hedging_statistics;
using volarb::InvalidHedgingInputsError;
using volarb::name_of_band_rule;
using volarb::whalley_wilmott_band;

namespace {

HedgingInputs base_inputs() {
    return HedgingInputs{100.0, 100.0, 0.25, 0.20, 63, 0.0010, 0.10};
}

const int paths = 400;

HedgingStatistics run(const BandPolicy& policy, const std::string& sequence) {
    return hedging_statistics(base_inputs(), policy, "1", sequence, paths);
}

}

TEST_CASE("the band scales as the cube root of cost and gamma squared", "[hedging]") {
    const double base = whalley_wilmott_band(100.0, 0.04, 0.001, 0.1);
    REQUIRE(whalley_wilmott_band(100.0, 0.04, 0.008, 0.1) == Approx(2.0 * base).epsilon(1e-12));
    REQUIRE(whalley_wilmott_band(100.0, 0.32, 0.001, 0.1) == Approx(4.0 * base).epsilon(1e-12));
    REQUIRE(whalley_wilmott_band(100.0, 0.04, 0.001, 0.8) == Approx(0.5 * base).epsilon(1e-12));
    REQUIRE(whalley_wilmott_band(100.0, 0.0, 0.001, 0.1) == Approx(0.0));
}

TEST_CASE("a wider band trades less and pays less", "[hedging]") {
    double previous_cost = 1e9;
    double previous_rebalances = 1e9;
    for (const double width : {0.0, 0.05, 0.10, 0.20, 0.40}) {
        const HedgingStatistics stats = run(BandPolicy{BandRule::Fixed, width}, "1");
        REQUIRE(stats.mean_transaction_cost < previous_cost);
        REQUIRE(stats.mean_rebalance_count < previous_rebalances);
        previous_cost = stats.mean_transaction_cost;
        previous_rebalances = stats.mean_rebalance_count;
    }
}

TEST_CASE("a wider band carries more risk", "[hedging]") {
    double previous = 0.0;
    for (const double width : {0.0, 0.05, 0.10, 0.20, 0.40}) {
        const HedgingStatistics stats = run(BandPolicy{BandRule::Fixed, width}, "1");
        REQUIRE(stats.profit_standard_deviation > previous);
        previous = stats.profit_standard_deviation;
    }
}

TEST_CASE("hedging with no cost and no band replicates the option", "[hedging]") {
    HedgingInputs inputs = base_inputs();
    inputs.proportional_cost = 0.0;
    inputs.steps = 252;
    const HedgingStatistics stats =
        hedging_statistics(inputs, BandPolicy{BandRule::Fixed, 0.0}, "1", "1", paths);
    REQUIRE(stats.mean_transaction_cost == Approx(0.0));
    REQUIRE(std::abs(stats.mean_profit) < 0.1);
    REQUIRE(stats.profit_standard_deviation < 0.3);
}

TEST_CASE("the same seed reproduces the same statistics", "[hedging]") {
    const HedgingStatistics one = run(BandPolicy{BandRule::WhalleyWilmott, 0.0}, "1");
    const HedgingStatistics other = run(BandPolicy{BandRule::WhalleyWilmott, 0.0}, "1");
    REQUIRE(one.mean_profit == other.mean_profit);
    REQUIRE(one.certainty_equivalent == other.certainty_equivalent);

    const HedgingStatistics elsewhere = run(BandPolicy{BandRule::WhalleyWilmott, 0.0}, "2");
    REQUIRE(elsewhere.mean_profit != one.mean_profit);
}

TEST_CASE("the certainty equivalent is the mean penalised by variance", "[hedging]") {
    const HedgingStatistics stats = run(BandPolicy{BandRule::Fixed, 0.10}, "1");
    const double variance = stats.profit_standard_deviation * stats.profit_standard_deviation;
    REQUIRE(stats.certainty_equivalent ==
            Approx(stats.mean_profit - 0.5 * base_inputs().risk_aversion * variance).epsilon(1e-12));
    REQUIRE(stats.certainty_equivalent < stats.mean_profit);
}

TEST_CASE("band rules have names that round trip", "[hedging]") {
    REQUIRE(band_rule_from_name("fixed") == BandRule::Fixed);
    REQUIRE(band_rule_from_name("whalley_wilmott") == BandRule::WhalleyWilmott);
    REQUIRE(name_of_band_rule(BandRule::Fixed) == "fixed");
    REQUIRE(name_of_band_rule(BandRule::WhalleyWilmott) == "whalley_wilmott");
    REQUIRE_THROWS_AS(band_rule_from_name("never"), InvalidHedgingInputsError);
}

TEST_CASE("malformed hedging inputs are rejected", "[hedging]") {
    HedgingInputs inputs = base_inputs();
    const BandPolicy policy{BandRule::Fixed, 0.1};
    inputs.spot = -1.0;
    REQUIRE_THROWS_AS(hedging_statistics(inputs, policy, "1", "1", paths), InvalidHedgingInputsError);
    inputs = base_inputs();
    inputs.risk_aversion = 0.0;
    REQUIRE_THROWS_AS(hedging_statistics(inputs, policy, "1", "1", paths), InvalidHedgingInputsError);
    inputs = base_inputs();
    REQUIRE_THROWS_AS(hedging_statistics(inputs, BandPolicy{BandRule::Fixed, -1.0}, "1", "1", paths),
                      InvalidHedgingInputsError);
    REQUIRE_THROWS_AS(hedging_statistics(inputs, policy, "1", "1", 1), InvalidHedgingInputsError);
    REQUIRE_THROWS_AS(hedging_statistics(inputs, policy, "x", "1", paths), InvalidHedgingInputsError);
}
